sled-agent/src/instance.rs

```diff
@@ -489,34 +488,23 @@ impl Instance {
         Ok(Instance { inner })
     }
 
-    fn create_opte_ports(
+    async fn setup_propolis_locked(
         &self,
         inner: &mut MutexGuard<'_, InstanceInner>,
-    ) -> Result<Vec<OptePort>, Error> {
-        let mut ports = Vec::with_capacity(inner.requested_nics.len());
+    ) -> Result<PropolisSetup, Error> {
+        // Create OPTE ports for the instance
+        let mut opte_ports = Vec::with_capacity(inner.requested_nics.len());
         for nic in inner.requested_nics.iter() {
-            let vni = crate::opte::Vni::new(nic.vni).expect("Invalid VNI");
             let external_ip =
                 if nic.primary { Some(inner.external_ip) } else { None };
-            let port = inner.port_allocator.new_port(
-                nic.ip,
-                *nic.mac,
-                ipnetwork::IpNetwork::from(nic.subnet),
-                vni,
-                inner.underlay_addr,
+            let port = inner.port_manager.create_port(
+                *inner.id(),
+                nic,
                 external_ip,
             )?;
-            info!(inner.log, "created OPTE port for guest"; "port_info" => ?port);
-            ports.push(port);
+            opte_ports.push(port);
         }
-        Ok(ports)
-    }
-
-    async fn setup_propolis_locked(
-        &self,
-        inner: &mut MutexGuard<'_, InstanceInner>,
-    ) -> Result<PropolisSetup, Error> {
-        let opte_ports = self.create_opte_ports(inner)?;
+        let port_ticket = opte_ports.first().map(|port| port.ticket());
```

Comment:
Because each ticket points to the whole `Vec<Port>`, so releasing one of them drops the whole vector and thus _all_ ports.

This is at least worth a comment, which I intended to do and forgot. I'm also open to suggestions here. The main issue is that there can be many ports for an instance. How do we release all of them, when we stop the instance? There are several options, such as this one, or having the key in the map be a tuple of the instance ID and port name. I'm not sure which is better, so I chose this. But if you've any ideas, I'm happy to hear them.

Reply:
I also realize now that this is a bit of an optimization that leaks the implementation details of the manager (that it stores things internally in a `BTreeMap<Uuid, Vec<Port>>`. Might be better to plug that leak and just store all tickets in the zone, even though the first one to call release will release all the others anyway.

Reply:
This reminds me of another thing I left out of the testing notes above. I forgot to include verification that this is in fact true, that keeping the first port ticket only, and calling `PortTicket::release` on it, correctly cleans up all ports.

I did a similar setup to the previous tests, but stopped an instance and added a secondary NIC to it, and started the instance again:

```
bnaecker@feldspar : ~/omicron $ oxide subnet create -o o -p p -v default -4 172.31.0.0/24 -D "second subnet" subnet2
✔ Created subnet subnet2 in o/p
bnaecker@feldspar : ~/omicron $ oxide api /organizations/o/projects/p/instances/i0/network-interfaces --method POST --input - <<EOF
{"name": "net1", "description": "net1", "vpc_name": "default", "subnet_name": "subnet2"}
EOF
{
  "description": "net1",
  "id": "7ff6f133-6210-46ff-87e6-109874f423cd",
  "instance_id": "41a3ef3c-ed12-48bf-93b8-155240f2448f",
  "ip": "172.31.0.5",
  "mac": "A8:40:25:FA:3E:DE",
  "name": "net1",
  "primary": false,
  "subnet_id": "ed4cd9b9-d6ff-4687-955d-7d97c779a148",
  "time_created": "2022-07-12T04:11:51.278853Z",
  "time_modified": "2022-07-12T04:11:51.278853Z",
  "vpc_id": "68d35ef2-e1fb-4c15-a384-1497381630c8"
}
bnaecker@feldspar : ~/omicron $ oxide instance start -o o -p p i0
✔  Waiting for instance status to be `running`
✔ Started instance i0 in o/p
```

At this point, we can see there are in fact two OPTE ports and two guest VNICs:

```
bnaecker@feldspar : ~/omicron $ dladm
LINK        CLASS     MTU    STATE    BRIDGE     OVER
igb0        phys      1500   up       --         --
net0        vnic      1500   up       --         igb0
net1        vnic      1500   up       --         igb0
stub0       etherstub 9000   up       --         --
underlay0   vnic      9000   up       --         stub0
oxControlService0 vnic 9000  up       --         stub0
oxControlStorage0 vnic 9000  up       --         stub0
oxControlStorage1 vnic 9000  up       --         stub0
oxControlStorage2 vnic 9000  up       --         stub0
oxControlStorage3 vnic 9000  up       --         stub0
oxControlStorage4 vnic 9000  up       --         stub0
oxControlService1 vnic 9000  up       --         stub0
oxControlService2 vnic 9000  up       --         stub0
oxControlPublic0 vnic 1500   up       --         igb0
oxControlService3 vnic 9000  up       --         stub0
opte1       xde       1500   up       --         --
vopte1      vnic      1500   up       --         opte1
opte2       xde       1500   up       --         --
vopte2      vnic      1500   up       --         opte2
oxControlInstance1 vnic 9000 up       --         stub0
bnaecker@feldspar : ~/omicron $ pfexec opteadm list-ports
LINK                             MAC ADDRESS              IPv4 ADDRESS     STATE
opte1                            A8:40:25:F4:67:FE        172.30.0.5       running
opte2                            A8:40:25:FA:3E:DE        172.31.0.5       running
bnaecker@feldspar : ~/omicron $ dladm show-linkprop -p secondary-macs net0
LINK         PROPERTY        PERM VALUE          DEFAULT        POSSIBLE
net0         secondary-macs  rw   a8:40:25:f4:67:fe --          --
bnaecker@feldspar : ~/omicron $
```

But you'll note there's only _one_ secondary MAC on `net0`. That's because only the primary NIC gets an external address and thus needs this proxy ARP trick. Also, stopping the instance removed both ports, and reset the linkprop:

```
bnaecker@feldspar : ~/omicron $ oxide instance stop -o o -p p i0
Type i0 to confirm stop:: i0
✔  Waiting for instance status to be `stopped`
✘ Stopped instance i0 in o/p
bnaecker@feldspar : ~/omicron $ dladm
LINK        CLASS     MTU    STATE    BRIDGE     OVER
igb0        phys      1500   up       --         --
net0        vnic      1500   up       --         igb0
net1        vnic      1500   up       --         igb0
stub0       etherstub 9000   up       --         --
underlay0   vnic      9000   up       --         stub0
oxControlService0 vnic 9000  up       --         stub0
oxControlStorage0 vnic 9000  up       --         stub0
oxControlStorage1 vnic 9000  up       --         stub0
oxControlStorage2 vnic 9000  up       --         stub0
oxControlStorage3 vnic 9000  up       --         stub0
oxControlStorage4 vnic 9000  up       --         stub0
oxControlService1 vnic 9000  up       --         stub0
oxControlService2 vnic 9000  up       --         stub0
oxControlPublic0 vnic 1500   up       --         igb0
oxControlService3 vnic 9000  up       --         stub0
bnaecker@feldspar : ~/omicron $ pfexec opteadm list-ports
LINK                             MAC ADDRESS              IPv4 ADDRESS     STATE
bnaecker@feldspar : ~/omicron $ dladm show-linkprop -p secondary-macs net0
LINK         PROPERTY        PERM VALUE          DEFAULT        POSSIBLE
net0         secondary-macs  rw   --             --             --
bnaecker@feldspar : ~/omicron $
```

Reply:
> I also realize now that this is a bit of an optimization that leaks the implementation details of the manager (that it stores things internally in a `BTreeMap<Uuid, Vec<Port>>`. Might be better to plug that leak and just store all tickets in the zone, even though the first one to call release will release all the others anyway.

Unless you think it would be totally unnecessary, this would be my inclination. If we build drop impls that are resistant to double calls, this seems like the least-likely-to-mess-up option if we start using multiple ports. Frankly, if I wrote this, my intuition would have been to use an `Arc` on the ports, to ensure that "if you have a reference to a port object, anywhere, that port is valid". However, that seems like it might be overkill, considering that we seem to only be referencing these ports on create / destroy.

Reply:
FWIW, my opinion on this is weakly held, and I do think at a minimum, your comment helps. So while I would personally store all the port tickets to hide that implementation detail, I trust your judgement.

Reply:
Sounds good. I think the approach you suggest is less magical, so I'll go ahead and store all tickets in the zone, and keep things in the manager internally in a different way to make that feasible.

> build drop impls that are resistant to double calls

On this point, it seems like the drop implementation is still resistant to double-calls, right? `impl Drop for PortTicket` calls `PortTicket::release`, which removes the port from the manager. If we call that twice, the removal would return `None`, since it's synchronized, no?

> my intuition would have been to use an Arc on the ports, to ensure that "if you have a reference to a port object, anywhere, that port is valid".

By "on the port" do you mean on the `Port` object, rather than internally on the `PortInner`, where the `Arc` currently is? It seems like the current implementation still achieves that aim, making sure that having a reference is only possible if the port is valid. Is there a way in which that can be subverted here?

Reply:
Storing each port independently is done in a98be984.